core/src/main/java/com/linecorp/armeria/common/ConnectionEventKey.java

```diff
@@ -0,0 +1,131 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import java.net.InetSocketAddress;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.util.AttributeKey;
+
+/**
+ * A key to identify a connection event.
+ */
+public final class ConnectionEventKey {
+    /**
+     * The attribute key of the {@link ConnectionEventKey}.
+     */
+    public static final AttributeKey<ConnectionEventKey> CONNECTION_EVENT_KEY =
+            AttributeKey.valueOf(ConnectionEventKey.class, "CONNECTION_EVENT_KEY");
+    /**
+     * The remote address of the connection.
+     */
+    private final InetSocketAddress remoteAddress;
+    /**
+     * The local address of the connection.
+     */
+    private final InetSocketAddress localAddress;
+    /**
+     * The protocol of the connection.
+     */
+    @Nullable
+    private SessionProtocol protocol;
+
+    /**
+     * The desired protocol of the connection. This field is only used for client-side connection.
+     */
+    @Nullable
+    private SessionProtocol desiredProtocol;
+
+    /**
+     * Whether the connection is active.
+     */
+    private boolean isActive;
+
+    /**
+     * Creates a new instance.
+     */
+    public ConnectionEventKey(InetSocketAddress remoteAddress,
+                              InetSocketAddress localAddress) {
+        this.remoteAddress = remoteAddress;
+        this.localAddress = localAddress;
+    }
+
+    /**
+     * Creates a new instance.
+     */
+    public ConnectionEventKey(InetSocketAddress remoteAddress,
+                              InetSocketAddress localAddress,
+                              SessionProtocol desiredProtocol) {
+        this.remoteAddress = remoteAddress;
+        this.localAddress = localAddress;
+        this.desiredProtocol = desiredProtocol;
+    }
+
+    /**
+     * Sets the protocol of the connection.
+     */
+    public ConnectionEventKey setProtocol(SessionProtocol protocol) {
+        this.protocol = protocol;
+        return this;
+    }
+
+    /**
+     * Sets whether the connection is active.
+     */
+    public ConnectionEventKey setActive(boolean isActive) {
+        this.isActive = isActive;
+        return this;
+    }
+
+    /**
+     * Returns the remote address of the connection.
+     */
+    public InetSocketAddress remoteAddress() {
+        return remoteAddress;
+    }
+
+    /**
+     * Returns the local address of the connection.
+     */
+    public InetSocketAddress localAddress() {
+        return localAddress;
+    }
+
+    /**
+     * Returns the protocol of the connection. This field can be null if the protocol is not determined yet.
+     */
+    @Nullable
+    public SessionProtocol protocol() {
+        return protocol;
+    }
+
+    /**
+     * Returns the desired protocol of the connection. This field is only used for client-side connection.
+     */
+    @Nullable
```

Comment:
I designed the ConnectionEventKey class (and ConnectionEventListener interface) for both the server and the client. Since the desired protocol only exists on the client side, this field has to be nullable for the server side.
Or perhaps we could extend this class to be more client-specific, but I'm getting worried that would make the code unnecessarily hierarchical and hard to read.